src/backend/distributed/planner/shard_pruning.c

```diff
@@ -578,25 +568,104 @@ PrunableExpressionsWalker(Node *node, ClauseWalkerContext *context)
 
 		return false;
 	}
-	else
+
+	return expression_tree_walker(node, PrunableExpressionsWalker, context);
+}
+
+
+/*
+ * AddSAOPartitionKeyRestrictionToInstance adds partcol = arrayelem operator
+ * restriction to the current pruning instance.
+ */
+static void
+AddSAOPartitionKeyRestrictionToInstance(ClauseWalkerContext *context,
+										ScalarArrayOpExpr *arrayOperatorExpression)
+{
+	PruningInstance *prune = context->currentPruningInstance;
+	Node *leftOpExpression = linitial(arrayOperatorExpression->args);
+	Node *strippedLeftOpExpression = strip_implicit_coercions(leftOpExpression);
+	bool usingEqualityOperator = OperatorImplementsEquality(
+		arrayOperatorExpression->opno);
+	Expr *arrayArgument = (Expr *) lsecond(arrayOperatorExpression->args);
+
+	/* found partcol = ANY(const, value, s); or partcol IN (const,b,c); */
+	if (usingEqualityOperator && strippedLeftOpExpression != NULL &&
+		equal(strippedLeftOpExpression, context->partitionColumn) &&
+		IsA(arrayArgument, Const))
 	{
-		PruningInstance *prune = context->currentPruningInstance;
+		ArrayType *array = NULL;
+		int16 typlen = 0;
+		bool typbyval = false;
+		char typalign = '\0';
+		Oid element_type = 0;
+		ArrayIterator arrayIterator = NULL;
+		Datum arg = 0;
+		Datum arrayDatum = ((Const *) arrayArgument)->constvalue;
```

Comment:
Naming here could be slightly more descriptive, e.g. `inArray` to convey that we're talking about the array of the `IN (..)`.